agent/eventhandler/task_handler.go

```diff
@@ -15,244 +15,323 @@ package eventhandler
-				done = true
-				return nil
-			}
+	delete(handler.tasksToEvents, taskARN)
+}
 
-			eventToSubmit := taskEvents.events.Front()
-			event := eventToSubmit.Value.(*sendableEvent)
-
-			if event.containerShouldBeSent() {
-				seelog.Infof("TaskHandler, Sending container change: %s", event.String())
-				err = client.SubmitContainerStateChange(event.containerChange)
-				if err == nil {
-					// submitted; ensure we don't retry it
-					event.setSent()
-					if event.containerChange.Container != nil {
-						event.containerChange.Container.SetSentStatus(event.containerChange.Status)
-					}
-					handler.stateSaver.Save()
-					seelog.Debugf("TaskHandler, Submitted container state change: %s", event.String())
-					backoff.Reset()
-					taskEvents.events.Remove(eventToSubmit)
-				} else {
-					seelog.Errorf("TaskHandler, Unretriable error submitting container state change [%s]: %v",
-						event.String(), err)
-				}
-			} else if event.taskShouldBeSent() {
-				seelog.Infof("TaskHandler, Sending task change: %s", event.String())
-				err = client.SubmitTaskStateChange(event.taskChange)
-				if err == nil {
-					// submitted or can't be retried; ensure we don't retry it
-					event.setSent()
-					if event.taskChange.Task != nil {
-						event.taskChange.Task.SetSentStatus(event.taskChange.Status)
-					}
-					handler.stateSaver.Save()
-					seelog.Debugf("TaskHandler, Submitted task state change: %s", event.String())
-					backoff.Reset()
-					taskEvents.events.Remove(eventToSubmit)
-				} else {
-					seelog.Errorf("TaskHandler, Unretriable error submitting task state change[%s]: %v",
-						event.String(), err)
-				}
-			} else if event.taskAttachmentShouldBeSent() {
-				seelog.Infof("TaskHandler, Sending task attachment change: %s", event.String())
-				err = client.SubmitTaskStateChange(event.taskChange)
-				if err == nil {
-					// submitted or can't be retried; ensure we don't retry it
-					event.setSent()
-					if event.taskChange.Attachment != nil {
-						event.taskChange.Attachment.SetSentStatus()
-						event.taskChange.Attachment.StopAckTimer()
-					}
-					handler.stateSaver.Save()
-					seelog.Debugf("TaskHandler, Submitted task attachment state change: %s", event.String())
-					backoff.Reset()
-					taskEvents.events.Remove(eventToSubmit)
-				} else {
-					seelog.Errorf("TaskHandler, Unretriable error submitting task attachment state change [%s]: %v",
-						event.String(), err)
-				}
-			} else {
-				// Shouldn't be sent as either a task or container change event; must have been already sent
-				seelog.Infof("TaskHandler, Not submitting redundant event; just removing: %s", event.String())
-				taskEvents.events.Remove(eventToSubmit)
-			}
+// sendChange adds the change to the sendable events list. If the
+// event for the task is in the process of being sent, it doesn't trigger
+// the async submit state api method
+func (taskEvents *taskSendableEvents) sendChange(change *sendableEvent,
```

Comment:
Can you add a bit to this comment here explaining that the actual task submission (API request) occurs asynchronously, so that this function doesn't end up causing the lock to be held for a long time?